File: chapter4/python/chapter4.py
def remove_blanks(str):
  index = 0
  while index < len(str):
    if str[index] == ' ':
      str = str[0 : index] + str[index + 1: :]
    else:
      index += 1
  return str

def get_digits(str):
  code = None
  result = ''
  for c in str:
    code= ord(c)
    #if (code >=  65 and code <= 90) ^ (code >= 97 and code <= 122):
    if code >= 48 and code <= 57:
      result += c
  return result

def acroynyms(str):
  state  = -1
  result = ''
  code = None

  for c in str:
    code = ord(c)
    if state == 0  and code == 32:
      state = -1

    if state == -1:
      if (code == 45) or  (code >=  65 and code <= 90) ^ (code >= 97 and code <= 122):
        state = 1

    if state == 1:
      result += c.capitalize()
      state = 0

  return result

def count_non_spaces (str):
  result = ''
  for c in str:
    if ord(c) is not 32:
      result += 1
  return result

def remove_shorter_string(str, length):
  words = str.split()
  result = ''
  for word in words:
    if len(word)  >=  length:
      result += word + ' '
  return result
      
def reverse_str(string):
  max = len(string)
  pos = 0
  rev_string= ''

  while pos < max:
    rev_string += string[max - pos - 1]
    pos += 1
  return rev_string 

def remove_even_strings(list_of_strings):
  index = 0
  while index < len(list_of_strings):
    if not (len(list_of_strings[index]) % 2):
      list_of_strings = list_of_strings[0 : index] + list_of_strings[index + 1: :]
    else:
      index += 1
  return list_of_strings




def integer_to_roman_numerals(number):
  numerals = {1: 'I', 5: 'V', 10: 'X', 50: 'L', 100: 'C', 500: 'D', 1000: 'M'}
  result = ''
  data = [] 
  list1 = sorted(numerals.keys())

  def double_digit_mode(a, number):
    data = None
    
    for num in list1: 
      if (num < a)  and (a - num <= number):
        data = [a, num] 
        break
    return data 

  while number > 0:
    data = [] 

    if number >= 1000:
      data = [1000]

    elif number >= 500:
      data = double_digit_mode(1000, number)
      data =  [500] if  data is None else data

    elif number >= 100:
      data = double_digit_mode(500, number)
      data =  [100] if data is None else data

    elif number >= 50: 
      data = double_digit_mode(100, number)
      data =  [50] if  data is None else data

    elif number >= 10:
      data = double_digit_mode(50, number)
      data =  [10] if  data is None else data

    elif number >= 5: 
      data = double_digit_mode(10, number)
      data =  [5] if data is None else data

    else:
      data = double_digit_mode(5, number)
      data =  [1] if data is None else data

    if len(data) == 2:
      result += numerals[data[1]] + numerals[data[0]]
      number -= data[0] - data[1]
    else:
      result += numerals[data[0]]
      number -= data[0]
    
  return result 

def roman_to_numerals (roman_numeral):
  numerals = {
    'I': 1, 
    'V': 5,
    'X': 10, 
    'L': 50,
    'C': 100,
    'D': 500,
    'M': 1000
  }
  
  index = 0
  result = ''
  length = len (roman_numeral) 
  sum = 0

  while index < length:
    curr = roman_numeral[index] 
    if index + 1 < length:
      next = roman_numeral[index + 1]        
    if next:
      if numerals[curr] < numerals[next]:
        sum +=  numerals[next] - numerals[curr]
        index += 2
        continue  
    sum += numerals[curr]
    index  += 1   
  return sum

def paren_valid(message):
  marker = { ')' : '('}
  open_brace = '('
  buffer = []

  for c in message:

    if c in open_brace:
      buffer.append(c)
    elif c in marker:
      if not len(buffer): 
        return False
      if buffer[-1] == marker[c]:
        buffer.pop()
      else:
        return False
  return True

def braces_valid(message):
  marker = {
    ')': '(',
    '}': '{',
    ']': '[',
    '<': '>',
  }
  opens = '({[<'
  buffer = [] 
  for c in message:

    if c in opens:
      buffer.append(c)
    elif c in marker:
      if not len(buffer): 
        return False
      if buffer[-1] == marker[c]:
        buffer.pop()
      else:
        return False
  return True

def is_palindrome(msg):
  index = 0 
  for i in range(len(msg) - 1, -1, -1):
    if msg[i] != msg[index]:
      return False 
    index += 1     
  return True

def is_palindrome_2(msg):
  index_left = 0
  index_right = len(msg) - 1
  
  while index_left <= index_right:

    if msg[index_left] == ' ':
      index_left += 1

    elif msg[index_right] == ' ':
      index_right -= 1
    
    else:

      if msg[index_left] != msg[index_right]:
        return False

      index_left += 1
      index_right -= 1
  return True

def longest_palindrome(msg):
  start_index = 0
  result = ''

  while start_index < len(msg):
    start_char = msg[start_index] 
    for end_index in range(len(msg) - 1, start_index - 1 , -1):
      end_char = msg [end_index] 
      if end_char == start_char:
        data = msg[start_index : end_index + 1]
        if is_palindrome(data):
          if len(result)  < len(data):
            result = data
    start_index += 1
  return result             

def is_word_alphabetical(msg):
  for i in range(1, len(msg)):
    if msg[i] != ' ' and msg[i-1] != ' ':
      if msg[i -1] > msg[i]:
        return False
  return True    

def d_gets_jiggy(name):
  if len(name) > 1:
    return name[1].capitalize() + name[2::] + ' ' + 'to the' + ' ' + name[0].capitalize() + '!'     
def common_suffix( word_list):
  result = ''
  index = 0 
 
  while True:
    for word in word_list:
      if index < len(word):
        c = word[len(word) -1 - index]
        if len(result) == index:
          result = c + result
        if result[0] != c :
          return result[1:]
      else:
        return result 
    index += 1

def book_index (pages): 
  buffer = []
  curr = []
  for i in range(0, len(pages) - 1):
    curr.append(pages[i])
    if pages[i] + 1 != pages[i + 1]:
      buffer.append(curr)
      curr = []
  if not curr:
    buffer.append([pages[-1]])
  return buffer    
  
 
def drop_the_mike(string):
  string = string.strip()
  result = ''

  for word in string.split():

    if word == 'Mike':
      return 'stunning violence'

    if len(word):
      a = word[0].capitalize() 
      b = word [1:] if len(word) > 1 else ''
      result += a + b + ' '
    
    return result 

def avg_of_obj(arr):
  obj = {
    'max': None,
    'min': None,
    'avg': 0
  }
  
  for num in arr:
    obj['avg'] += num 
  
    if obj['max'] == None or num > max:
      obj['max'] = num
 
    if obj['min'] == None or num < min:
      obj['min'] = num

  if len(arr):
    obj['avg'] = obj['avg'] / len(arr)  

  return obj  

def zip_arrays_to_map(a,b):
  result = {}

  if len(a) != len(b):
    return None
  
  for i in range(0, len(a)):
     result [a[i]] = b[i]
  
  return result 

def invert_hash(obj):
  for key in obj:
    obj[obj[key]] = key
    del obj[key] 
  return obj

def number_of_kv(obj):
  count = 0
  for ele in obj:
    count += 1
  return count
 
x =number_of_kv( {'od': 22, 'paul': 'moon'  } )
print(x)

File: chapter4/java/chapter4.java
package algo;
import java.math.*;
import java.util.*;
import java.util.Map.Entry;

public class Chapter4 {
	
	static String removeBlank (String str) 
	{
		String s = "";
		
		for (String retVal: str.split(" "))
		{
			s += retVal;
		}
		return s;
	}
	
	static Integer getDigits (String str) 
	{
		String s = "";
		
		for (String retVal: str.split("")) 
		{
			if ( retVal.codePointAt(0) >= 48 && retVal.codePointAt(0) <= 57  )  // is number !
			{
				s += retVal;
			}
		}
		return Integer.valueOf(s);
	}
	
	static String acronyms(String str)
	{
		String s = "";
		for (String retVal: str.split(" "))
		{
			if ( Character.isLetter(retVal.codePointAt(0)))
			{
				s += Character.toUpperCase( retVal.charAt(0) );
			}
		}
		return s;
	}
	
	static int countNonSpaces (String str) 
	{
		int count = 0;
		
		for (String retVal: str.split("")) 
		{
			if (retVal.codePointAt(0) != 0x20)
			{
				count++;
			}
		}
		return count;
	}
	
	static String removeShorterString (String str, int len) 
	{
		String s = "";
		for (String retVal: str.split(" ")) 
		{
			if (retVal.length() >= len) 
			{
				s += retVal + " ";
			}
		}
		return s;
	}
	
	static String reverseString(String  str)
	{
		String s ="";
		int left, right;
		
		left =  (int )Math.ceil(str.length() / 2 ) - 1 ; 
		right = left + 1;

		while (true) 
		{				
			if (left >= 0) 
			{
				s += str.charAt( left);				
				left--;
			}

			if (right < str.length()) 
			{
				s = str.charAt(right) + s;
				right++;
			}

			if (left < 0 && right >= str.length()) 
			{
				break;
			}
		}
		return s;
	}
	
	static void removeEvenString(ArrayList<String>  collection)
	{
		for ( int i = 0; i < collection.size(); ) 
		{
			if ((collection.get(i).length() % 2) == 0) // even ! 
			{
				collection.remove(i);
			}
			else 
			{
				i++;
			}
		}
	}
	
	static String integerToRomanNumeral (int number)
	{
		String s = "";
		int[] sel = new int[2];
		Hashtable< Integer, String> table = new Hashtable<Integer,String >();
		
		//set table
		table.put( 0	,"");
		table.put( 1	,"I");
		table.put( 5	,"V");
		table.put( 10	,"X");
		table.put( 50	,"L");
		table.put( 100	,"C");
		table.put( 500	,"D");
		table.put( 1000	,"M"); 
		
		while (number > 0)
		{

			switch( number )
			// fall through 
			{

			default:

			case 1000:
				
				if (number >= 900 && number <= 1000)  // two char mode  [900 - 1000]
				{
					sel[1] = 100;
				}
				
				if (number >= 1000 || sel[1]  == 100) 
				{
					sel[0]  = 1000;
					break;
				}
				
			case 500:
				
				if (number >= 400 && number <= 500)  // two char mode [400 - 500]
				{
					sel[1] = 100;
				}
				
				if (number >= 500 || sel[1] == 100) 
				{
					sel[0]  = 500;
					break;
				}
			case 100:
				
				if (number >= 90 && number <= 100)  // two char mode  [90 - 100]
				{
					sel[1] = 10;
				}
				
				if (number >= 100 || sel[1] == 10) 
				{
					sel[0]  = 100;
					break;
				}
			case 50:
				
				if (number >= 40 && number <= 50)  // two char mode   [40 - 50]
				{
					sel[1] = 10;
				}
				
				if (number >= 50 || sel[1] == 10) 
				{
					sel[0]  = 50;
					break;
				}
				
			case 10:
				
				if (number >= 9 && number <= 10)  // two char mode  [9 - 10]
				{
					sel[1] = 1;
				}
				
				if (number >= 10 || sel[1] == 1) 
				{
					sel[0]  = 10;
					break;
				}
				
			case 5:
				
				if (number >= 4 && number <= 5)  // two char mode [4 - 5]
				{
					sel[1] = 1;
				}
				
				if (number >= 5 || sel[1] == 1) 
				{
					sel[0]  = 5;
					break;
				}
				
			case 1:
				
				if (number >= 1) 
				{
					sel[0]  = 1;
				}		
			}
		
			number -= (sel[0]  - sel[1]);
			s += table.get(sel[1]) + table.get(sel[0] );
			sel[0] = sel[1] = 0;
		}
		
		return s;
	}
	
	static int romanNumeralToInteger(String romanNumeral)
	{
		int result = 0;
		Hashtable< Character, Integer> table = new Hashtable< Character, Integer>();
		int[] sel = new int[2];
		int i = 0;
		
		//set table
		table.put( 	'I', 1		);
		table.put( 	'V', 5		);
		table.put( 	'X', 10		);
		table.put( 	'L', 50		);
		table.put( 	'C', 100	);
		table.put( 	'D', 500	);
		table.put( 	'M', 1000	); 
		
		while (i < romanNumeral.length())
		{
			sel[0] = table.get(romanNumeral.charAt(i)) ;
			
			if (i + 1 < romanNumeral.length())
			{
				sel[1] = table.get(romanNumeral.charAt(i + 1));
			}
			
			if (sel[0] < sel[1]) 
			{
				i += 2;
				result += sel[1] - sel[0];
			}
			else 
			{
				result += sel[0];
				i++;
			}
		}
		
		return result;
	}
	
	static boolean parensValid(String str) 
	{
		String s = "";
		int counter = 0;
		
		for (String retVal: str.split("") ) 
		{
			if (retVal.charAt(0) == '(') 
			{
				counter++;
			}
			
			if (retVal.charAt(0) == ')') 
			{
				if (counter == 0) 
				{
					return false;
				}
				counter--;
			}
		}
		return counter == 0;
	}
	
	static boolean bracesValid (String str) 
	{
		Hashtable<Character, Character> table = new Hashtable<Character, Character>();
		ArrayList<Character> array = new ArrayList<Character>(); // front = back of list 
		String openBraces = "([{";
		Character c;
		
		table.put(')', '(');
		table.put('}', '{');
		table.put(']', '[');
			
		for (String retVal: str.split("")) 
		{
			c = retVal.charAt(0);
			
			if ( openBraces.indexOf(c) != -1) // open brace
			{
				array.add(0, c);
			}

			if (table.containsKey(c)) // close brace
			{

				if (array.size() == 0) // array empty!
				{
					return false;
				}
				if (table.get(c) != array.get(0)) // not a match!
				{
					return false;
				}
				
				array.remove(0);
			}
		}
		return array.size() == 0; // buffer must be empty 
	}
	
	/* strict palindrome */
	static boolean isPalindrome (String str)   
	{
		int halfSize = str.length() / 2;
		
		for (int i = 0; i < halfSize; i++) 
		{
			if (str.charAt(i) != str.charAt(  str.length() - 1 - i) )
			{
				return false;
			}
		}
		return true;
	}
	
	/*  palindrome evaluates numbers and alphabets  */
	static boolean isAlphaNumericPalindrome (String str)   
	{
		Character c0, c1;
		boolean[] isAlphaNumericChar = new boolean[2];
		int[] index = {0, str.length() - 1};
		
		while ( index[0] <= index[1]) 
		{
			c0 = str.charAt(index[0]);
			c1 = str.charAt(index[1]);
			isAlphaNumericChar[0] = Character.isAlphabetic(c0) || Character.isDigit(c0);
			isAlphaNumericChar[1] = Character.isAlphabetic(c1) || Character.isDigit(c1);
			
			if (isAlphaNumericChar[1] && isAlphaNumericChar[0] ) // headptr and tailptr are alpha numeric 
			{
				if (c0 != c1) 
				{
					return false;
				}
			}
			
			if (!isAlphaNumericChar[0])
			{
				index[0]++;
				continue;
			}
			
			if (!isAlphaNumericChar[1])
			{
				index[1]--;
				continue;
			}
			
			index[0]++;
			index[1]--;
		}
		
	
		return true;
	}
	
	static String longestPalindrome(String str) 
	{
		String palindrome = "";
		String subs;
		int[] index = new int[2];
		
		for (int i = 0; i < str.length(); i++)
		{
			index[0] = i;
			index[1] = str.length() - 1;

			while ( index[0] <= index[1]) 
			{
				subs = str.substring(index[0], index[1] + 1);

				if ( subs.charAt(0) == subs.charAt(subs.length() - 1) ) // possible palindrome  x[0] = x[end]
				{
					if (subs.length() > palindrome.length()) 
					{
						if (isPalindrome(subs))  // try isAlphaNumericPalindrome() !
						{
							palindrome = str.substring(index[0], index[1] + 1);							
							break;						
						}
					}
				}
				index[1]--;					
			}
		}
		
		return palindrome;
	}
	
	static String dGetsJiggy(String name) 
	{
		String s = "";
		Character c, cend;
		String msg = "to the";
		
		if (name.length() <= 0) 
		{
			return "";
		}
		
		c = Character.toUpperCase(name.charAt(0));
		
		if (name.length() == 1) 
		{
			return  msg + " " + c + "!";
		}
		
		if (name.length() > 1) 
		{
			return Character.toUpperCase(name.charAt(1)) +  name.substring(2, name.length()) + " "  + msg + " " + c + "!";
		}
		
		return "";
	}

	static boolean isWordAlphabetical (String str) 
	{
		int index = 0;
		Character c;
		Character[] data = new Character[2];

		while (index < str.length())
		{
			c = str.charAt(index);
			
			if ( Character.isAlphabetic(c) )
			{
				data[0] = data[1];
				data[1] = c;
			}
			
			if (data[0] !=null && data[1] !=null) 
			{
				if (data[0] > data[1]) 
				{
					return false;
				}
			}
			
			index++;					
		}
		return true;
	}
	
	static String commonSuffix (ArrayList<String> words) 
	{
		
		String  curr, next, result ="";
		int lenFromEnd = 0; 
		boolean done = false;
		
		if (words.size() == 0) 
		{
			return "";
		}
		
		while (!done) 
		{
			lenFromEnd++;

			for (int i = 0; i < words.size() - 1; i++)  
			{
				curr = words.get(i);
				next = words.get(i + 1);
				
				if ( !(lenFromEnd <= curr.length() && lenFromEnd <= next.length() ) ) // window greater than word length  
				{
					done = true;
				}
				else if ( curr.charAt(curr.length()  - lenFromEnd) !=  next.charAt(next.length()  - lenFromEnd) ) // adjacent elements don't match
				{
					done = true;
				}
				
				if (done) 
				{
					lenFromEnd--; // decr 1 step from error position 
					break;
				}
			}
		}
		
		if (lenFromEnd > 0) 
		{
			curr = words.get(0);
			result = curr.substring(curr.length() - lenFromEnd, curr.length());
		}
		
		return result;
	}
	
	static String bookIndex (int[] numbers)
	{
		String s = "";
		Integer start = null, curr, end = null, prev;
		
		if (numbers.length == 0) 
		{
			return s;
		}
		
		prev = start = (numbers[0]);
				
		for (int i = 1; i < numbers.length; i++) 
		{
			curr = numbers[i];
			
			if (prev + 1 != curr)  // discontinuity
			{
				if (s.length() != 0) 
				{
					s += ", ";
				}
				
				if (start != null) 
				{
					s += String.valueOf(start);
				}
				
				if (end != null) 
				{
					s +=  '-' + String.valueOf(end) ;
				}	
				
				// replace start
				start = curr; 
				end = null;
			}
			
			else 
			{
				end = curr;
			}
			prev = curr;
		}
		
		if (start != null) 
		{
			s += ", ";
			s += String.valueOf(start);
		}
	
		if (end != null) 
		{
			s +=  '-' + String.valueOf(end) ;
		}	
		
		return s;
	}
	
	public static void main (String agrs[]) 
	{
		int[] array = {1, 13, 14, 15, 37, 38, 70};
		Object s = bookIndex(array);
		System.out.print(s);
	}
}
